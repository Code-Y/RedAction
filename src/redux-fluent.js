export { default as createAction } from './Action/Action';

export { default as createReducer } from './Reducer/Reducer';

export { default as createCombinableReducers } from './CombinableReducers/CombinableReducers';
