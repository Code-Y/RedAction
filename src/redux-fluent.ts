export { createAction } from './create-action/create-action';
export { createReducer } from './create-reducer/create-reducer';
export { ofType, RFR } from './of-type/of-type';
