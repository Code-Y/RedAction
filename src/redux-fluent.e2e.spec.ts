import { combineReducers, createStore } from 'redux';
import { isError } from 'flux-standard-action';
import { createAction, createReducer, ofType } from './redux-fluent';

interface Todo {
  id: string;
  completed: boolean;
}

const actions = {
  greet: createAction<'messages | greet'>('messages | greet'),
  addTodo: createAction<'todos | add', Todo>('todos | add'),
  deleteTodo: createAction<'todos/:id | delete', Todo['id'], void, Todo>(
    'todos/:id | delete',
    (todo) => todo.id,
  ),
};

const todos = createReducer<'todos', Todo[]>('todos')
  .actions(
    ofType(actions.addTodo).map((state, action) =>
      state.concat(action.payload),
    ),
    ofType(actions.deleteTodo).map((state, action) =>
      isError(action) ? state : state.filter(({ id }) => id !== action.payload),
    ),
  )
  .default(() => []);

const errors = createReducer<'errors', string[]>('errors')
  .actions(
    ofType(actions.addTodo).map((state, action) =>
      isError(action) ? [...state, action.payload.message] : state,
    ),
  )
  .default(() => []);

describe('redux-fluent E2E', () => {
  it('should add a todo', () => {
    const reducer = combineReducers({ todos, errors });
    const store = createStore(reducer);

    const todo = {
      id: 'first-todo-id',
      completed: false,
    };

    store.dispatch(actions.addTodo(todo));
    expect(store.getState()).toMatchSnapshot();
  });

  it('should delete a todo', () => {
    const reducer = combineReducers({ todos, errors });
    const store = createStore(reducer);

    const todo = {
      id: 'first-todo-id',
      completed: true,
    };

    store.dispatch(actions.addTodo(todo));
    expect(store.getState()).toMatchSnapshot();

    store.dispatch(actions.deleteTodo(new Error('something went wrong')));
    expect(store.getState()).toMatchSnapshot();

    store.dispatch(actions.deleteTodo(todo));
    expect(store.getState()).toMatchSnapshot();
  });
});
