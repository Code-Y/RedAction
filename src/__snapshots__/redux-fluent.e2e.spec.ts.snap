// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux-fluent E2E should add a todo 1`] = `
Object {
  "todos": Array [
    Object {
      "id": "first-todo-id",
      "status": "pending",
    },
  ],
}
`;

exports[`redux-fluent E2E should not add a todo if action.error 1`] = `
Object {
  "todos": Array [],
}
`;
